pkg/kubelet/apis/config/validation/validation_linux.go

```diff
@@ -28,9 +28,14 @@ import (
 
 // validateKubeletOSConfiguration validates os specific kubelet configuration and returns an error if it is invalid.
 func validateKubeletOSConfiguration(kc *kubeletconfig.KubeletConfiguration) error {
-	if kc.FailCgroupV1 && !libcontainercgroups.IsCgroup2UnifiedMode() {
+	isCgroup1 := !libcontainercgroups.IsCgroup2UnifiedMode()
+	if kc.FailCgroupV1 && isCgroup1 {
 		return fmt.Errorf("kubelet is configured to not run on a host using cgroup v1. cgroup v1 support is in maintenance mode")
 	}
 
+	if isCgroup1 && kc.SingleProcessOOMKill != nil && !*kc.SingleProcessOOMKill {
+		return fmt.Errorf("invalid configuration: singleProcessOOMKill must not be explicitly set to false when using cgroup v1")
+	}
+
```

Comment:
Similar to the windows case.  If we use kubeadm to install on nodes among which some are using cgroup v1 and some are using cgroup v2, the same configuration cannot apply to both as it will fail on cgroup v1 node. This will make it more complex to manage a cluster like that.


If we want to fail on cgroup v1, we can use `FailCgroupV1`.

https://github.com/kubernetes/kubernetes/blob/7b802a60bd6e43ab53499e3b0a19c06f395974df/pkg/kubelet/apis/config/types.go#L498-L503

If the node is using cgroup v1, we can just ignore the configuration with a warning.

Reply:
FailSwapOn is another example. kubelet exit during running(not during configuration validation) when FailSwapOn is true and the node swap is enabled.

Reply:
@roycaihw @SergeyKanzhelev WDYT?

Reply:
Ignore with the warning is OK with me.

Reply:
sounds good to me too

Reply:
I think we are doing a tradeoff between 
- easy-to-use (in some cases) API, v.s.
- comprehensive and reliable API

If we drop the validation, and do warning + internal override, one downside is that it could be misleading when user troubleshoots node behavior. For example, a user may wrongly expect that a node is doing group oom kill based on `singleProcessOOMKill=false`, without knowing that the node is using cgroup v1 and the config is internally overridden to `true`.

I'm in favor of keeping the validation. WDYT?

(On the other hand, if we decide to go with the easy-to-use direction, we should be consistent between Windows and Linux, meaning we would also just warning for Windows)

Reply:
I am not against failing kubelet if `singleProcessOOMKill` setting cannot work on the node.

Take `FailCgroupV1` as an example: FailCgroupV1 means failing kubelet on cgroup v1. However, this is not a configuration validation error. It is a running error.

It means, even we only log a warning during validation, we can fail the kubelet during running/starting. 

To gain the similar behavior, you can choose to fail or not if we just warning here.
```
failCgroupV1: true
singleProcessOOMKill: true
```
this means the kubelet will fail if it runs on cgroup v1.

```
failCgroupV1: false
singleProcessOOMKill: true
```
this means the kubelet will just warning if it runs on cgroup v1.

Similarly, if the cluster has windows and Linux nodes and using the same kubelet configuration, the windows node would not fail and you can fail it with another configuration.

Reply:
@kubernetes/api-reviewers PTAL.

Reply:
> For example, a user may wrongly expect that a node is doing group oom kill based on `singleProcessOOMKill=false`, without knowing that the node is using cgroup v1 and the config is internally overridden to `true`.

For cgroup version, we can check the metrics kubelet_cgroup_version in case user want to check if singleProcessOOMKill is really enabled.
```
[root@paco ~]#  kubectl get --raw /api/v1/nodes/$(kubectl get nodes --no-headers | awk '{print $1}')/proxy/metrics  | grep kubelet_cgroup_version
# HELP kubelet_cgroup_version [ALPHA] cgroup version on the hosts.
# TYPE kubelet_cgroup_version gauge
kubelet_cgroup_version 2
```

Reply:
SGTM. @utam0k Could you address @pacoxu's other comment https://github.com/kubernetes/kubernetes/pull/126096#discussion_r1804239114 to also warning on Windows, so that the behavior is consistent?

Reply:
I'm a little confused about the validation v.s. warning behavior, and would like to get some clarification @liggitt @SergeyKanzhelev @pacoxu. My understanding is that there are two direction, and currently the PR seems to be mixing both. Am I missing something? @liggitt IIUC you were suggesting Direction 1 below. Could you clarify? 

---

Current PR

| Machine spec                          | User input (SingleProcessOOMKill)  | Erroring v.s. warning |
|---------------------------------------|------------------------------------|-----------------------|
| cgroup v1  (with failCgroupV1: false) | false                              | Invalid               |
| Windows                               | true / false                       | Warning               |
| Other OS                              | true / false                       | Invalid               |

---

Direction 1: Validate and error on unsupported settings.

| Machine spec                          | User input (SingleProcessOOMKill)  | Erroring v.s. warning |
|---------------------------------------|------------------------------------|-----------------------|
| cgroup v1  (with failCgroupV1: false) | false                              | Invalid               |
| Windows                               | true / false                       | Invalid               |
| Other OS                              | true / false                       | Invalid               |

--- 

Direction 2: Warning only
| Machine spec                          | User input (SingleProcessOOMKill)  | Erroring v.s. warning        |
|---------------------------------------|------------------------------------|------------------------------|
| cgroup v1  (with failCgroupV1: false) | false                              | Warning and treating as true |
| Windows                               | true / false                       | Warning                      |
| Other OS                              | true / false                       | Warning                      |

Reply:
Correct, the direction 1 table is what I expect.

The config file fields are nil by default. Specifying a non-nil value requires the value to be valid for the platform and configuration the user is on.

Reply:
https://github.com/kubernetes/kubernetes/pull/126096#discussion_r1825368347 Thanks for clarifying @liggitt. 

I think we are still doing a [tradeoff](https://github.com/kubernetes/kubernetes/pull/126096#discussion_r1809721476) regarding the [two directions](https://github.com/kubernetes/kubernetes/pull/126096#discussion_r1821663977). @pacoxu @utam0k @marosset @SergeyKanzhelev I wonder how we can reach consensus here. WDYT?

Reply:
@liggitt just to make sure I understand the logic. Is the Windows is a special case when we use warnings instead of failing for settings that cannot be applied? My original suggestion here was to fail everywhere. But if we special case Windows, because it is hard to configre kubeadm for it, why not special case the cgroupv1 as well? From my perspective it will be as hard if not harder to apply different configs based on which cgroup is used

Reply:
so I feel either we are being strict all over and not allow to set those for neither Windows nor cgroupv1. Or we just warn in both cases

Reply:
Synced with @liggitt @pacoxu @SergeyKanzhelev @utam0k offline. The consensus was that the validation will fail all unsupported `SingleProcessOOMKill` configs

Reply:
+1